common/lib/dependabot/new_version.rb

```diff
@@ -0,0 +1,228 @@
+# typed: strong
+# frozen_string_literal: true
+
+require "sorbet-runtime"
+require "dependabot/version"
+
+module Dependabot
```

Comment:
Most of this code is from `maven/version.rb`. As explained in the PR description, this is one of the most comprehensive versioning implementations we have and I've decided to build on it for this task. 

It will eventually be renamed to `version.rb`. I've had to add it because my approach is incremental so all ecosystems other than maven will still use the old version.rb and I'll incrementally subclass from `new_version.rb` until all ecosystems are covered.

Reply:
A lot of the code that's extracted into this new class seems very specific to Maven, for example the prefix tokens or named qualifiers are not valid in for example npm or Bundler:

<img width="434" alt="Screenshot 2024-08-06 at 15 54 17" src="https://github.com/user-attachments/assets/cd001359-9c2d-4108-8684-3b88a05282b5">

```ruby
irb(main):001> Gem::Version.new("1.0.0_pre1")
/Users/jurre/.rbenv/versions/3.3.1/lib/ruby/site_ruby/3.3.0/rubygems/version.rb:223:in `initialize': Malformed version number string 1.0.0_pre1 (ArgumentError)

      raise ArgumentError, "Malformed version number string #{version}"
```

It probably makes sense to keep things that are specific to the maven ecosystem in the maven implementation, and to only extract those things that are generic

Reply:
The goal is to standardize how dependabot handles versioning so we have an opinionated approach for all ecosystems. We do see non-standard version strings in the npm logs and having a standardized approach helps us to reason about the space. 

[SemVer2 actually supports pre-release fields and this is a step in that direction](https://semver.org/).

Reply:
Maven's version spec is not compatible with SemVer 2: https://maven.apache.org/pom.html#version-order-specification

> The Maven version order algorithm is not compatible with Semantic Versioning 2.0.0. In particular, Maven does not special case the plus sign or consider build identifiers.

Each ecosystem diverges from SemVer in some way, I don't know of one that perfectly matches the spec. Currently we base common's version on Gem::Version which is also a variant of SemVer, and each ecosystem diverges from Gem::Version in some way or another as captured by their particular implementation.

So I'm not sure what swapping out the base Version class from Gem's variant to Maven's variant will achieve. If we wanted to reimplement the base Version to be exactly to spec, that might make more sense!

> The goal is to standardize how dependabot handles versioning so we have an opinionated approach for all ecosystems.

I don't see how we can be opinionated about SemVer in our implementation, we're trying to exactly copy each ecosystem's behavior in Ruby. Any variance results in a bug report that we're not matching native behavior. 

The best approach is probably have our base Version satisfy the SemVer spec exactly, and then each ecosystem only implements the variance. Again, this is kind of what we do now but the variance is from Gem::Version and not the SemVer spec.